dsmr/node.go

```diff
@@ -0,0 +1,247 @@
+			sk:       sk,
+			verifier: chunkVerifier,
+			storage:  storage,
+		},
+		ChunkCertificateGossipHandler: &ChunkCertificateGossipHandler[T]{
+			storage: storage,
+		},
+		storage: storage,
+		pk:      bls.PublicFromSecretKey(sk),
+	}, nil
+}
+
+type Node[T Tx] struct {
+	nodeID                       ids.NodeID
+	sk                           *bls.SecretKey
+	networkID                    uint32
+	chainID                      ids.ID
+	getChunkClient               *TypedClient[*dsmr.GetChunkRequest, *dsmr.GetChunkResponse, []byte]
+	getChunkSignatureClient      *TypedClient[*dsmr.GetChunkSignatureRequest, *dsmr.GetChunkSignatureResponse, []byte]
+	chunkCertificateGossipClient *TypedClient[[]byte, []byte, *dsmr.ChunkCertificateGossip]
+	validators                   []Validator
+
+	GetChunkHandler               *GetChunkHandler[T]
+	GetChunkSignatureHandler      *GetChunkSignatureHandler[T]
+	ChunkCertificateGossipHandler *ChunkCertificateGossipHandler[T]
+	storage                       *chunkStorage[T]
+	pk                            *bls.PublicKey
+}
+
+// NewChunk builds transactions into a Chunk
+// TODO handle frozen sponsor + validator assignments
+func (n *Node[T]) NewChunk(
+	ctx context.Context,
+	txs []T,
+	expiry int64,
+	beneficiary codec.Address,
+) (Chunk[T], error) {
+	if len(txs) == 0 {
+		return Chunk[T]{}, ErrEmptyChunk
+	}
+
+	chunk, err := signChunk[T](
+		UnsignedChunk[T]{
+			Producer:    n.nodeID,
+			Beneficiary: beneficiary,
+			Expiry:      expiry,
+			Txs:         txs,
+		},
+		n.sk,
+		n.pk,
+		n.networkID,
+		n.chainID,
+	)
+	if err != nil {
+		return Chunk[T]{}, fmt.Errorf("failed to sign chunk: %w", err)
+	}
+
+	protoChunk, err := newProtoChunk[T](chunk)
+	if err != nil {
+		return Chunk[T]{}, fmt.Errorf("failed to marshal chunk: %w", err)
+	}
+
+	request := &dsmr.GetChunkSignatureRequest{
+		Chunk: protoChunk,
+	}
+
+	for _, validator := range n.validators {
+		done := make(chan struct{})
+		onResponse := func(context.Context, ids.NodeID, *dsmr.GetChunkSignatureResponse, error) {
+			defer close(done)
+			// TODO generate chunk cert
+		}
+
+		if err := n.getChunkSignatureClient.AppRequest(ctx, validator.NodeID, request, onResponse); err != nil {
+			return Chunk[T]{}, err
+		}
+
+		<-done
+	}
+
+	chunkCert := &ChunkCertificate{
+		ChunkID:   chunk.id,
+		Expiry:    chunk.Expiry,
+		Signature: NoVerifyChunkSignature{},
+	}
+
+	if err := n.chunkCertificateGossipClient.AppGossip(ctx, &dsmr.ChunkCertificateGossip{
+		ChunkCertificate: &dsmr.ChunkCertificate{
+			ChunkId:   chunk.id[:],
+			Producer:  chunk.Producer[:],
+			Expiry:    chunk.Expiry,
+			Signers:   nil, // TODO populate
+			Signature: nil,
+		},
+	}); err != nil {
+		return Chunk[T]{}, err
+	}
+
+	return chunk, n.storage.AddLocalChunkWithCert(chunk, chunkCert)
+}
+
+func (n *Node[T]) NewBlock(parent Block, timestamp int64) (Block, error) {
+	if timestamp <= parent.Timestamp {
+		return Block{}, ErrTimestampNotMonotonicallyIncreasing
+	}
```

Comment:
We may want to allow them to be equal to handle the case a malicious peer sets their timestamp as far ahead as possible, we can either handle that by failing to build a block or allowing duplicate timestamps.

Reply:
`NewBlock` should only be called from code building blocks which will be local, so I don't think we have to worry about a malicious timestamp being passed in